File: src/code-samples/langchain/streaming-agent-progress.py
# :snippet-start: streaming-agent-progress-py
from langchain.agents import create_agent
from langchain_core.utils.uuid import uuid7
from langgraph.checkpoint.memory import InMemorySaver

def get_weather(city: str) -> str:
    """Get weather for a given city."""
    return f"It's always sunny in {city}!"

agent = create_agent(
    model="gpt-5-nano",
    tools=[get_weather],
    checkpointer=InMemorySaver()
)
config = {"configurable": {"thread_id": str(uuid7())}}
stream = agent.stream_events(  # [!code highlight]
    {"messages": [{"role": "user", "content": "What is the weather in SF?"}]},
    config=config,
    version="v3",  # [!code highlight]
)
for snapshot in stream.values:  # [!code highlight]
    print(f"content: {snapshot['messages'][-1].content_blocks}")
# :snippet-end:

# :remove-start:
if __name__ == "__main__":
    stream = agent.stream_events(
        {"messages": [{"role": "user", "content": "What is the weather in SF?"}]},
        config={"configurable": {"thread_id": str(uuid7())}},
        version="v3",
    )
    snapshots = list(stream.values)
    stream.output
    assert len(snapshots) > 0, snapshots
    print("✓ streaming agent progress (stream_events v3) emits value snapshots")
# :remove-end:

File: src/code-samples/langchain/streaming-agent-progress.ts
// :snippet-start: streaming-agent-progress-js
import { createAgent, tool } from "langchain";
import { MemorySaver } from "@langchain/langgraph";
import z from "zod";

const getWeather = tool(
  async ({ city }) => {
    return `The weather in ${city} is always sunny!`;
  },
  {
    name: "get_weather",
    description: "Get weather for a given city.",
    schema: z.object({
      city: z.string(),
    }),
  },
);

const agent = createAgent({
  model: "gpt-5-nano",
  tools: [getWeather],
  checkpointer: new MemorySaver(),
});

const config = { configurable: { thread_id: crypto.randomUUID() } };

const stream = await agent.streamEvents(
  { messages: [{ role: "user", content: "what is the weather in sf" }] },
  { ...config, version: "v3" },
);
for await (const snapshot of stream.values) {
  console.log(
    `content: ${JSON.stringify(snapshot.messages.at(-1).contentBlocks, null, 2)}`,
  );
}
// :snippet-end:

// :remove-start:
async function main() {
  const collected: unknown[] = [];
  const stream = await agent.streamEvents(
    { messages: [{ role: "user", content: "what is the weather in sf" }] },
    {
      configurable: { thread_id: crypto.randomUUID() },
      version: "v3",
    },
  );
  await Promise.all([
    (async () => {
      for await (const snapshot of stream.values) {
        collected.push(snapshot);
      }
    })(),
    stream.output,
  ]);
  if (collected.length === 0) {
    throw new Error("expected at least one stream values snapshot");
  }
  console.log(
    "✓ streaming agent progress (streamEvents v3) emits value snapshots",
  );
}

main();
// :remove-end:
